src/librustc/middle/const_eval.rs

```diff
@@ -913,12 +937,16 @@ pub fn eval_const_expr_with_substs<'tcx, S>(tcx: &ty::ctxt<'tcx>,
                           def::FromTrait(trait_id) => match tcx.map.find(def_id.node) {
                               Some(ast_map::NodeTraitItem(ti)) => match ti.node {
                                   ast::ConstTraitItem(ref ty, _) => {
-                                      let substs = get_substs(e.id);
-                                      (resolve_trait_associated_const(tcx,
-                                                                      ti,
-                                                                      trait_id,
-                                                                      substs),
-                                       Some(&**ty))
+                                      if let ExprTypeChecked = ty_hint {
+                                          let substs = tcx.node_id_item_substs(e.id).substs;
```

Comment:
(It took me a little while to convince myself that this ends up being equivalent to the effect we were getting before via the `get_substs` parameter, but I now believe the tables in the `tcx` should match even in the case where `check::_match::check_pat` was passing in a closure that calls `fcx.item_substs()`, so this should be okay.)